libgenic/power.c

```diff
@@ -75,33 +97,49 @@ void parse_power(int i, double k, char * line, struct table *out_tab, int * Inpu
+    k -= log10(scale);  /* convert to h/Kpc */
     out_tab->logk[i] = k;
-    /*CDM, Baryons, photons, massless nu, massive nu, tot, CDM+bar, other stuff*/
-    for(j = 0; j< 7; j++) {
+    /* Note: the ncdm entries change depending on the number of neutrino species. The first row, k,
+     * is read in read_power_table and passed as a parameter.
+     * but only the first is used for particles.
+     * 1:k (h/Mpc)              2:d_g                    3:d_b                    4:d_cdm                  5:d_ur
+     * 6:d_ncdm[0]              7:d_ncdm[1]              8:d_ncdm[2]              9:d_tot                 10:phi
+     * 11:psi                   12:h                     13:h_prime               14:eta                   15:eta_prime
+     * 16:t_g                   17:t_b                   18:t_ur
+     * 19:t_ncdm[0]             20:t_ncdm[1]             21:t_ncdm[2]             22:t_tot*/
+    for(j = 0; j< ncols; j++) {
         char * retval = strtok(NULL, " \t");
+        /*This happens if we do not have as many neutrino species as we expect, or we don't find h_prime.*/
         if(!retval)
-            endrun(1,"Incomplete line in power spectrum: %s\n",line);
+            endrun(1,"Incomplete line in power spectrum: only %d columns, expecting %d. Did you remember to set extra metric transfer functions=y?\n",j, ncols);
         transfers[j] = atof(retval);
     }
     /*Order of the transfer table matches the particle types:
-     * 0 is baryons, 1 is CDM, 2 is CMB + baryons, 3 is massive neutrinos.
-     * Everything is a ratio against tot.
-     * In the input CAMB file 0 is CDM, 1 is baryons, 2 is photons,
-     * 3 is massless neutrinos, 4 is massive neutrinos, 5 is total,
-     * 6 is cdm + baryons.*/
-    out_tab->logD[0][i] = transfers[1]/transfers[5];
-    out_tab->logD[1][i] = transfers[0]/transfers[5];
-    out_tab->logD[2][i] = transfers[4]/transfers[5];
-    out_tab->logD[3][i] = transfers[6]/transfers[5];
+     * 0 is baryons, 1 is CDM, 2 is massive neutrinos (we use the most massive neutrino species).
+     * 3 is growth function for baryons, 4 is growth function for CDM, 5 is growth function for massive neutrinos.
+     * We use the formulae for velocities from fastpm in synchronous gauge:
+     * https://github.com/rainwoodman/fastpm-python/blob/master/fastpm/multi.py#L185
+     * CDM = - h_prime / 2 / d_cdm
+     * bar = - (h_prime / 2  + t_b) / d_b
+     * nu = - (h_prime / 2 + t_ncdm) / d_ncdm
+     * and there is a normalisation factor of (1+z)/ hubble applied later on.*/
+    out_tab->logD[0][i] = -1*transfers[1];
+    out_tab->logD[1][i] = -1*transfers[2];
+    /*This is ur if neutrinos are massless*/
+    out_tab->logD[2][i] = -1*transfers[3+nnu];
+    /*h_prime is entry 8 + nnu. t_b is 12 + nnu, t_ncdm[2] is 13 + nnu * 2.*/
+    out_tab->logD[3][i] = transfers[12+nnu];
+    out_tab->logD[4][i] = transfers[8+nnu] * 0.5;
+    out_tab->logD[5][i] = transfers[13+nnu*2];
```

Comment:
The unit here is in CLASS  units. (time derivative is against CLASS unit of time). Could you add a xref to the function that does the unit conversion to GADGET unit?